android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java

```diff
@@ -462,29 +462,102 @@ public static boolean executeHardwareKeystroke(
   }
 
   /**
-   * Handle the globe key longpress action. Currently just for INAPP only
+   * Handle the globe key action on the lock screen
+   * @param context
+   */
+  private static void doGlobeKeyLockscreenAction(Context context) {
+    // Override globe key action if screen is locked:
+    // 1. Switch to next Keyman keyboard (no menu)
+    // 2. When all the Keyman keyboards have been cycled through, advance to the next system keyboard
+    if (sysKbStartingIndexOnLockScreen == getCurrentKeyboardIndex(context)) {
+      // all the Keyman keyboards have been cycled through
+      advanceToNextInputMode();
+    } else {
+      if (sysKbStartingIndexOnLockScreen == -1) {
+        // Initialize the system keyboard starting index while the screen is locked
+        sysKbStartingIndexOnLockScreen = getCurrentKeyboardIndex(context);
+      }
+      switchToNextKeyboard(context);
+    }
+  }
+
+  /**
+   * Handle the globe key longpress action.
    * @param context
    * @param keyboard
    */
   private static void doGlobeKeyLongpressAction(Context context, KeyboardType keyboard) {
-    if (keyboard == KeyboardType.KEYBOARD_TYPE_INAPP && InAppKeyboard != null) {
-      if (InAppKeyboard.keyboardPickerEnabled) {
-        showKeyboardPicker(context, keyboard);
+    boolean keyboardPickerEnabled = false;
+    if (keyboard == KeyboardType.KEYBOARD_TYPE_INAPP) {
+      if (InAppKeyboard == null) {
+        return;
       }
+      keyboardPickerEnabled = InAppKeyboard.keyboardPickerEnabled;
+    } else if (keyboard == KeyboardType.KEYBOARD_TYPE_SYSTEM) {
+      if (SystemKeyboard == null) {
+        return;
+      }
+      keyboardPickerEnabled = SystemKeyboard.keyboardPickerEnabled;
+    }
+
+    if (keyboardPickerEnabled) {
+      showKeyboardPicker(context, keyboard);
     }
   }
 
   /**
-   * Handle the globe key shortpress action. Currently just for INAPP only
+   * Handle the globe key shortpress action
    * @param context
-   * @param keyboard
+   * @param keyboard KeboardType of KEYBOARD_TYPE_INAPP or KEYBOARD_TYPE_SYSTEM
    */
   private static void doGlobeKeyShortpressAction(Context context, KeyboardType keyboard) {
-    if (keyboard == KeyboardType.KEYBOARD_TYPE_INAPP && InAppKeyboard != null) {
-      if (InAppKeyboard.keyboardPickerEnabled && inappKbGlobeKeyAction == GlobeKeyAction.GLOBE_KEY_ACTION_SHOW_MENU) {
-        showKeyboardPicker(context, KeyboardType.KEYBOARD_TYPE_INAPP);
-      } else { /* inappKbGlobeKeyAction == GlobeKeyAction.GLOBE_KEY_ACTION_SWITCH_TO_NEXT_KEYBOARD */
-        switchToNextKeyboard(context);
+    GlobeKeyAction action = GlobeKeyAction.GLOBE_KEY_ACTION_DO_NOTHING;
+    boolean keyboardPickerEnabled = false;
```

Comment:
This pattern is similar to the previous function. We probably don't really need this variable at all; it would be better to treat the `keyboardPickerEnabled` as an input condition and then short-circuit return, see below: